### ISSUE TITLE
Authorization required for authorization_code grant

### ISSUE THREAD

**Issue by JoeKolba** [slavab89/oauth2-server-example-mongodb#2]
I am trying to make a request to /oauth/authorize with the client_id, client_secret, grant_type, response_type and redirect_uri but it keeps throwing a 401 unauthorized error.  I shoudn't need to authenticate in order to test this should I?

**Comment by slavab89** [comment]
What's the exact message that you get back from the server? It can help track down the exact place that is failing.
Also, what are you returning [oauthServerMiddlewares](https://github.com/slavab89/oauth2-server-example-mongodb/blob/master/oauthServerMiddlewares.js#L13)? 
See https://github.com/oauthjs/node-oauth2-server/issues/358 as well.
If i'm right you are not returning a valid user object so you're getting a 401 error back.

Basically, OAuth2 only handle authorization for a service, it does not handle authentication (user login).

**Comment by dongmai** [comment]
@slavab89 , Can you tell me if my request is wrong? 

`http://localhost:8000/oauth/authorize?client_id={client_id}&client_secret={client_secret}&response_type=code&redirect_uri=http://localhost:8000/`

It always fails as it keeps returning:

`{"message":"Unauthorized request: no authentication given","error":{"statusCode":401,"status":401,"code":401,"message":"Unauthorized request: no authentication given","name":"unauthorized_request"}}`

Thanks

**Comment by slavab89** [comment]
Your request seems to be correct except 2 things compared to what i did:
1. i have an additional `state` query param (Just set it to some random string)
2. `Authorization` header with a bearer access token.

Regarding point 2, and the issue that i've previously written.
You might be asking yourself: "I want to get an authorization_code in order to get an access_token, how can i possibly have a access_token? Where do i get this access_token, that will let me get an authorization_code, that will let me get another access_token???"
This is the exact question that i've asked myself and what the previously linked issue is talking about.

You can get that access_token using one of 2 methods:

- grant type: password
- grant type: client_credentials

For grant type password (POST request):
You should send `Authorization` header with client_id and client_secret hashed in base64 format.
You should send a body with `grant_type` `password` and a username and password fields.

For grant type client_credentials (Post request):
You should send a body with `grant_type` `client_credentials` and a client_id and client_secret fields.

One of those methods will get you an access_token that you will be able to use for the authorization_code flow.
I'm not sure how to build the whole flow and where to place the user authentication as i've decided to use OpenId Connect instead of OAuth2.

**Comment by slavab89** [comment]
I've just added a few examples of the different requests that you can do

**Comment by dongmai** [comment]
I got this, Thank @slavab89 💯
